lib/src/main/java/io/ably/lib/http/Http.java

```diff
@@ -272,62 +312,31 @@ private static AblyException checkFor(Throwable t) {
 
 			/* send request body */
 			if(requestBody != null) {
-				conn.setDoOutput(true);
-				byte[] body = requestBody.getEncoded();
-				int length = body.length;
-				conn.setFixedLengthStreamingMode(length);
-				conn.setRequestProperty(CONTENT_TYPE, requestBody.getContentType());
-				conn.setRequestProperty(CONTENT_LENGTH, Integer.toString(length));
-				OutputStream os = conn.getOutputStream();
-				os.write(body);
+				writeRequestBody(requestBody, conn);
 			}
 
-			/* get response */
-			statusCode = conn.getResponseCode();
-			statusLine = conn.getResponseMessage();
-			if(statusCode != HttpURLConnection.HTTP_NO_CONTENT) {
-				contentType = conn.getContentType();
-				int contentLength = conn.getContentLength();
-				int successStatusCode = (method == POST) ? HttpURLConnection.HTTP_CREATED : HttpURLConnection.HTTP_OK;
-				is = (statusCode == successStatusCode) ? conn.getInputStream() : conn.getErrorStream();
-				if(is != null) {
-					int read;
-					if(contentLength == -1) {
-						ByteArrayOutputStream baos = new ByteArrayOutputStream();
-						byte[] buf = new byte[4 * 1024];
-						while((read = is.read(buf)) > -1) {
-							baos.write(buf, 0, read);
-						}
-						responseBody = baos.toByteArray();
-					} else {
-						int idx = 0;
-						responseBody = new byte[contentLength];
-						while((read = is.read(responseBody,  idx, contentLength - idx)) > -1) {
-							idx += read;
-						}
-					}
-				}
-			}
+			response = readResponse(conn);
 		} catch(IOException ioe) {
 			throw HostFailedException.checkFor(ioe);
 		} finally {
-			try {
-				if(is != null)
-					is.close();
-				if(conn != null)
-					conn.disconnect();
-			} catch(IOException ioe) {}
+			if(conn != null)
+				conn.disconnect();
 		}
 
-		if(statusCode == 0) {
+		if (response.statusCode == 0) {
```

Comment:
I don't think the HTTP request will ever validly return a `statusCode` of 0. That is, it would only happen if the server really did put a `0` in the status line of the response. If it is unable to parse the status line the returned code will be -1.

I think the correct behaviour there is to throw an exception signifying an invalid response if the `statusCode` is not in the range 2xx-5xx.

The previous code looks like it was checking for 0 as a way of deciding whether or not there had been any response, but the correct action in that case would have been to throw instead of silently returning `null`.